src/Elmish.WPF/Merge.fs

```diff
@@ -40,9 +40,9 @@ module CollectionTarget =
       Move = oc.Move
       Clear = oc.Clear
       Enumerate = fun () -> upcast oc
-      BoxedCollection = fun () -> oc |> box }
+      GetCollection = fun () -> oc }
 
-  let map (fOut: 'a -> 'b) (fIn: 'b -> 'a) (ct: CollectionTarget<'a>) : CollectionTarget<'b> =
+  let map (fOut: 'a0 -> 'a1) (fIn: 'a1 -> 'a0) (ct: CollectionTarget<'a0, 'aCollection>) : CollectionTarget<'a1, 'aCollection> =
```

Comment:
It doesn't make sense to me how `'a0` can change to `'a1` but `'aCollection` doesn't change.

Reply:
We're just mapping the element type not the collection type. For example, `IEnumerable<int>` and `IEnumerable<obj>` both implement `IEnumerable`. It's kind of the same idea.

We maintain knowledge of the collection type even as we erase knowledge of the element type. To do this completely generically, we have to map them separately.

Reply:
In this particular case, we are boxing up the element types but keeping knowledge of the collection type (because that ultimately is the type of the property). We can't keep the element types around statically afaik and we don't really need to.

Reply:
> We're just mapping the element type not the collection type.

I don't think so.

The only calls to `CollectionType.map` are immediately followed by `CollectionType.mapCollection`.  Going up the (static) call stack, these are all implementations of `boxMinorTypes`, which maps `'a` to `obj` and `ObservableCollection<'a>` to `obj`.

I think it would make more sense if `CollectionType` had a single `map` function with this signature:

```fsharp
  let map (outMapA: 'a -> 'a0)
          (outMapACollection: 'aCollection -> 'a0Collection0)
          (inMapA: 'a0 -> 'a)
          (ct: CollectionTarget<'a, 'aCollection>) =
```

Reply:
@TysonMN so the very next pull request splits these out. I'd rather do this here than there. If you want I can shuffle it, but it will immediately get shuffled back to doing it the way it is done here in #470.

See [this diff between PR 470 and this PR](https://github.com/marner2/Elmish.WPF/compare/feature/add_collection_type_bindings...marner2:Elmish.WPF:feature/add_vm_type_information). I'm moving `CollectionTarget.mapCollection` out of `boxMinorTypes` and into `BindingData.boxT`.

Reply:
Let's shuffle it.  Then the next PR usage shuttle justify them being separate.

Reply:
Done and shuffled